protocol/contracts/templegold/SpiceAuction.sol

```diff
@@ -0,0 +1,253 @@
+        id += 1;
+        SpiceAuctionConfig storage config = auctionConfigs[id];
+        if (config.duration == 0) { revert InvalidConfigOperation(); }
+        delete auctionConfigs[id];
+        /// @dev `_currentEpochId` is only updated when auction starts
+        emit AuctionConfigRemoved(id);
+    }
+
+    /**
+     * @notice Start auction. Checks caller is set config starter. Address zero for anyone to call
+     */
+    function startAuction() external override {
+        uint256 epochId = _currentEpochId;
+        /// @dev config is always set for next auction
+        /// @notice Configuration is set before auctions so configId = currentEpochId + 1;
+        SpiceAuctionConfig storage config = auctionConfigs[epochId+1];
+        if (config.duration == 0) { revert CannotStartAuction(); }
+        /// @notice only starter
+        if (config.starter != address(0) && msg.sender != config.starter) { revert CommonEventsAndErrors.InvalidAccess(); }
+        /// @notice enough wait period since last auction
+        if (epochId > 0) {
+            /// @dev `_currentEpochId` is still last epoch
+            EpochInfo memory lastEpochInfo = epochs[epochId];
+            /// use waitperiod from last auction config
+            uint64 _waitPeriod = auctionConfigs[epochId].waitPeriod;
+            if (lastEpochInfo.endTime + _waitPeriod > block.timestamp) { revert CannotStartAuction(); }
+        } else {
+            /// For first auction
+            if (_deployTimestamp + config.waitPeriod > block.timestamp) { revert CannotStartAuction(); }
+        }
+        address auctionToken = _getAuctionTokenForEpochStorage(config);
+        uint256 totalAuctionTokenAllocation = _totalAuctionTokenAllocation[auctionToken];
+        uint256 balance = IERC20(auctionToken).balanceOf(address(this));
+        uint256 epochAuctionTokenAmount = balance - (totalAuctionTokenAllocation - _claimedAuctionTokens[auctionToken]);
+        if (config.activationMode == ActivationMode.AUCTION_TOKEN_BALANCE) {
+            if (config.minimumDistributedAuctionToken == 0) { revert MissingAuctionTokenConfig(); }
+        }
+        if (epochAuctionTokenAmount < config.minimumDistributedAuctionToken) { revert NotEnoughAuctionTokens(); }
+        // epoch start settings
+        // now update currentEpochId
+        epochId = _currentEpochId = _currentEpochId + 1;
+        EpochInfo storage info = epochs[epochId];
+        uint64 startTime = info.startTime = uint64(block.timestamp) + config.startCooldown;
+        uint64 endTime = info.endTime = uint64(block.timestamp) + config.duration;
+        info.totalAuctionTokenAmount = epochAuctionTokenAmount;
+
+        // Keep track of total allocation auction tokens per epoch
+        _totalAuctionTokenAllocation[auctionToken] = totalAuctionTokenAllocation + epochAuctionTokenAmount;
+
+        emit AuctionStarted(epochId, msg.sender, startTime, endTime, epochAuctionTokenAmount);
+    }
+
```

Comment:
What's the flow for the auction tokens to get into this contract?

Would it be sensible to add a `recoverToken()` to this contract (as well as the others)? Eg if a partner accidentally sends way to much and wants to recover some before the auction starts.

If the auction has already started, and trying to recover the spiceToken then it should revert. But otherwise i think fair game?

And/or what if you need to migrate to a new SpicAuction for the same token (eg new logic). There could be locked tokens?

Reply:
partner simply sends auction tokens to contract, and before auction starts, we check tokens sent are at least `minimumDistributedAuctionToken`.

Sure, will add `recoverToken()` 👍

Good call. For migration,  user can still claim retro with the same (old) spice contract. But the new auction which is not started will be deleted and the tokens sent back to `recipient`. And then the tokens can be sent to new spice contract. So it's either:
1/ `function migrateNextAuctionTokens(address _recipient) external;`
2/ `function migrateNextAuctionTokens(ISpiceAuction _newSpiceContract) external onlyMigrator`.  where `onlyMigrator` means function is called by new spice auction contract.
What do you think?

Reply:
```
    function recoverToken(uint256 amount, address to) external onlyDAOExecutor {
        if (to == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }

        uint256 epochId = _currentEpochId;
        EpochInfo storage info = epochs[epochId];
        if (info.isActive()) { InvalidConfigOperation(); }
        if (info.startTime == 0) { InvalidConfigOperation(); }

        /// @dev could be that `auctionStart` is triggered but there's cooldown, which is not reached (so can delete epochInfo for _currentEpochId)
        // or `auctionStart` is not triggered but `auctionConfig` is set (where _currentEpochId is not updated yet)
        
        if (info.startTime - lastAuctionTriggerTimestamp != config.startCooldown) {
            // `auctionStart` is not triggered but `auctionConfig` is set
            epochId += 1;
        }

        SpiceAuctionConfig storage config = auctionConfigs[epochId];
        if (config.duration == 0) { revert InvalidConfigOperation(); }

        address auctionToken = _getAuctionTokenForEpochStorage(config);
        uint256 totalAuctionTokenAllocation = _totalAuctionTokenAllocation[auctionToken];
        uint256 balance = IERC20(auctionToken).balanceOf(address(this));
        uint256 epochAuctionTokenAmount = balance - (totalAuctionTokenAllocation - _claimedAuctionTokens[auctionToken]);
        if (amount > epochAuctionTokenAmount) { revert CommonEventsAndErrors.InvalidParam(); }
        uint256 amountRemaining;
        unchecked {
            amountRemaining = epochAuctionTokenAmount - amount;
        }
        if (amountRemaining < config.minimumDistributedAuctionToken) { revert NotEnoughAuctionTokens();  }
        IERC20(auctionToken).safeTransfer(to, amount);
    }

```

Reply:
1/ This handles the auction token, what about other tokens sent to the auction? 
-- The bid token
-- Any other erc20 - should still be able to recover those without any checks?

2/ 
For the case where the auction has been started but still in cooldown, could it be simplified by enforcing that removeAuction() has to be called first for that case? So then within this function instead of:
`if (info.isActive()) { InvalidConfigOperation(); }`

can use:
`if (!info.hasEnded()) { InvalidConfigOperation(); }` // Or another error like 'StillActive'

Reply:
In terms of migration, my vote would be to keep it quite straight forward.
1/ Surplus auction tokens can be recovered using this function. Now this contract is only usable for users to claim historic auction tokens
2/ Create a new auction contract and set it up fresh, with a fresh auction config set and epoch starts at zero again.

I don't think that needs a migration function?

Reply:
migration: I agree to 1/ and 2/

Reply:
updated `recoverToken()`:

```
    function recoverToken(address token, address to, uint256 amount) external onlyDAOExecutor {
        if (to == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
        if (amount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }

        if (token != spiceToken && token != templeGold) {
            emit CommonEventsAndErrors.TokenRecovered(to, token, amount);
            IERC20(token).safeTransfer(to, amount);
            return;
        }

        uint256 epochId = _currentEpochId;
        EpochInfo storage info = epochs[epochId];
        /// @dev use `removeAuctionConfig` for case where `auctionStart` is called and cooldown is still pending
        if (!info.hasEnded()) { revert StillActive(); }
        if (auctionConfigs[epochId+1].duration == 0) { revert RemoveAuctionConfig(); }

        /// @dev Now `auctionStart` is not triggered but `auctionConfig` is set
    
        // check to not take away intended tokens for claims
        // calculate auction token amount
        uint256 totalAuctionTokenAllocation = _totalAuctionTokenAllocation[token];
        uint256 balance = IERC20(token).balanceOf(address(this));
        uint256 maxRecoverAmount = balance - (totalAuctionTokenAllocation - _claimedAuctionTokens[token]);
        
        if (amount > maxRecoverAmount) { revert CommonEventsAndErrors.InvalidParam(); }
        
        IERC20(token).safeTransfer(to, amount);

        emit CommonEventsAndErrors.TokenRecovered(to, token, amount);
    }
```